pip/pip-330.md

```diff
@@ -0,0 +1,43 @@
+# PIP-330: getMessageById gets all messages
+
+# Motivation
+
+The `org.apache.pulsar.client.admin.Topics` provides `getMessageById(java.lang.String, long, long)` method to get the
+message, which returns one message. If the message id refers to a batch message, we only can get the first message, not
+all messages.
+
+This behavior affects our analysis of messages by the message id.
+
+# Goals
+
+## In Scope
+
+Add a method that returns all messages by message id to the `org.apache.pulsar.client.admin.Topics` interface.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+Add `getMessageById(java.lang.String, java.lang.String)` to `org.apache.pulsar.client.admin.Topics` interface:
+
+```java
+public interface Topics {
+    List<Message<byte[]>> getMessageById(String topic, String messageId) throws PulsarAdminException;
```

Comment:
Can you please make the `String messageId` not a string, and break it down as instructed below to ledgerId, entryId and batchIndex?

Reply:
It seems like you can retain the exact signature we have today:
```
  public Message<byte[]> getMessageById(String topic, long ledgerId, long entryId)
```
and perhaps change the name to signify you will get *all* messages:

```
  public Message<byte[]> getAllMessagesById(String topic, long ledgerId, long entryId)
```
and you can add to get a specific one, by the batch index:
```
  public Message<byte[]> getMessageById(String topic, long ledgerId, long entryId, long batchIndex)
```

Reply:
> It seems like you can retain the exact signature we have today:
> 
> ```
>   public Message<byte[]> getMessageById(String topic, long ledgerId, long entryId)
> ```

I should add `@Deprecate` for `public Message<byte[]> getMessageById(String topic, long ledgerId, long entryId)`.

We need to consider the batch message(msg1, msg2, msg3), when you use the method, you can only get the `msg1`, not all. This will create a false impression that the user thinks there is only one message.



> and perhaps change the name to signify you will get _all_ messages:
> 
> ```
>   public Message<byte[]> getAllMessagesById(String topic, long ledgerId, long entryId)
> ```
> 
> and you can add to get a specific one, by the batch index:
> 
> ```
>   public Message<byte[]> getMessageById(String topic, long ledgerId, long entryId, long batchIndex)
> ```

I prefer to `public Message<byte[]> getMessageById(String topic, long ledgerId, long entryId, long batchIndex)`, when batchIndex is greater than -1, we can extract the specific one from the batch message by batchIndex.

The current discussion point is whether we should use a `String` parameter or multiple parameters(`long ledgerId, long entryId, long batchIndex`). IMO, everything is fine.

Reply:
HI, @nodece, my original motivation is that we should hide the internals like ledger id. However, as an administrator, he/she has to know these details anyway. Even if we provide an API that accepts a string, it must be documented to include some fields as well, no matter it's a JSON (`{"ledger": 10, "entry": 20}` or a comma-separated string (`10:20`).

So I agree with @asafm's suggestion that we should not use the string parameter.

And I also agree with you that the original API should be deprecated. Getting only 1 message looks strange.

Reply:
Thanks @asafm and @BewareMyPower, this PIP has been updated, could you review this PR again?